clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -293,7 +293,9 @@ public static DefaultRecord readFrom(ByteBuffer buffer,
                                          Long logAppendTime) {
         int sizeOfBodyInBytes = ByteUtils.readVarint(buffer);
         if (buffer.remaining() < sizeOfBodyInBytes)
-            return null;
+            throw new InvalidRecordException("Invalid record size: expected " + sizeOfBodyInBytes +
+                " bytes in record payload, but instead the buffer has only " + buffer.remaining() +
+                " remaining bytes.");
```

Comment:
Is this really an exceptional case? Don't we do reads where we don't know exactly where the read ends and hence will trigger this path?

Reply:
Are you saying the case that we are yet to complete reading the request? I didn't see a retry path, but it will cause a null point exception at https://github.com/apache/kafka/blob/bfc57aa4ddcd719fc4a646c2ac09d4979c076455/core/src/main/scala/kafka/log/LogValidator.scala#L191

What do you suggest me do here?

Reply:
I think the intent here was to cover the case where an incomplete record is returned by the broker. However, we have broker logic to try and avoid this case since KIP-74:

```java
} else if (!hardMaxBytesLimit && readInfo.fetchedData.firstEntryIncomplete) {
            // For FetchRequest version 3, we replace incomplete message sets with an empty one as consumers can make
            // progress in such cases and don't need to report a `RecordTooLargeException`
            FetchDataInfo(readInfo.fetchedData.fetchOffsetMetadata, MemoryRecords.EMPTY)
```

@hachikuji Do you remember if there is still a reason to return `null` here instead of the exception @ccding is proposing?

Reply:
> the case where an incomplete record is returned by the broker

I am referring to the produce API in this PR. The record is from producers.

Reply:
Yes, I understand you're talking about the producer case. I am talking about the fetch case. As I said, I think we may not need that special logic anymore, but @hachikuji would know for sure.

Reply:
@hachikuji do you have time to have a look at this?

Reply:
Apologies for the delay here. I don't see a problem with the change. I believe that @ijuma is right that the fetch response may still return incomplete data, but I believe this is handled in `ByteBufferLogInputStream`. We stop batch iteration early if there is incomplete data, so we would never reach the `readFrom` here which is called for each record in the batch. It's worth noting also that the only caller of this method (in `DefaultRecordBatch.uncompressedIterator`) has the following logic:

```java
                try {
                    return DefaultRecord.readFrom(buffer, baseOffset, firstTimestamp, baseSequence, logAppendTime);
                } catch (BufferUnderflowException e) {
                    throw new InvalidRecordException("Incorrect declared batch size, premature EOF reached");
                }
```

So it is already handle underflows in a similar way.

Reply:
Thanks for checking @hachikuji.